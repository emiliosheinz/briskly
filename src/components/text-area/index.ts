export { TextArea } from './text-area.component'
