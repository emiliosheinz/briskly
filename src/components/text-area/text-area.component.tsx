import React from 'react'

import { classNames } from '~/utils/css'

import { Tooltip } from '../tooltip'
import type { TextAreaProps } from './text-area.types'

export const TextArea = React.forwardRef<HTMLTextAreaElement, TextAreaProps>(
  function TextArea(props, ref) {
    const {
      label,
      error,
      id,
      disabled,
      rows = 5,
      tooltip,
      ...otherProps
    } = props

    const errorClassNames = error
      ? 'border-error-700 focus:border-error-700 focus:ring-error-700'
      : 'border-primary-500 focus:border-primary-900 focus:ring-primary-900'

    return (
      <div
        aria-disabled={disabled}
        className={classNames('w-full', disabled ? 'opacity-50' : '')}
      >
        <span className='flex'>
          <label
            htmlFor={id}
            className='block text-sm font-medium capitalize text-primary-800'
          >
            {label}
          </label>
          {tooltip ? <Tooltip hint={tooltip} /> : null}
        </span>
        <div className='mt-1 rounded-md shadow-sm'>
          <textarea
            id={id}
            ref={ref}
            rows={rows}
            disabled={disabled}
            data-testid='briskly-custom-textarea'
            className={classNames(
              'block w-full rounded-md bg-primary-50 pl-6 pr-10 text-base text-primary-900',
              errorClassNames,
            )}
            {...otherProps}
          />
        </div>
        <p className='mt-1 h-5 overflow-hidden text-ellipsis whitespace-nowrap text-sm text-error-700'>
          {error}
        </p>
      </div>
    )
  },
)
