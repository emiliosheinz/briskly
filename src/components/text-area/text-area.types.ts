export type TextAreaProps = {
  error?: string
  label: string
  id: string
  tooltip?: string
} & Omit<
  React.DetailedHTMLProps<
    React.TextareaHTMLAttributes<HTMLTextAreaElement>,
    HTMLTextAreaElement
  >,
  'id'
>
