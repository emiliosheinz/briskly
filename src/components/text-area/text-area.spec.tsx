import { render, screen } from '~/utils/tests/react-testing-library'

import { TextArea } from './text-area.component'

describe('TextArea Component', () => {
  it('should be correctly rendered', () => {
    const component = render(<TextArea id='mock' label='Label' />)

    expect(component.asFragment()).toMatchSnapshot()
  })

  it('should be correctly rendered when hasError', () => {
    const component = render(
      <TextArea id='mock' label='Label' error='Mocked Error' />,
    )

    expect(component.asFragment()).toMatchSnapshot()
  })

  it('should be correctly rendered when disabled', () => {
    const component = render(<TextArea disabled id='mock' label='Label' />)

    expect(component.asFragment()).toMatchSnapshot()
  })

  it('should show input label', () => {
    render(<TextArea id='mock' label='Label' />)

    const label = screen.getByText('Label')

    expect(label).toBeInTheDocument()
  })

  it('should show input error', () => {
    render(<TextArea id='mock' label='Label' error='Mocked Error' />)

    const error = screen.getByText('Mocked Error')

    expect(error).toBeInTheDocument()
  })

  it('should disable input when disabled is provided', () => {
    render(<TextArea disabled id='mock' label='Label' error='Mocked Error' />)

    const input = screen.getByTestId('briskly-custom-textarea')

    expect(input).toBeDisabled()
  })

  it('should have 5 rows as default', () => {
    render(<TextArea id='mock' label='Label' />)

    const input = screen.getByTestId('briskly-custom-textarea')

    expect(input.getAttribute('rows')).toBe('5')
  })
})
