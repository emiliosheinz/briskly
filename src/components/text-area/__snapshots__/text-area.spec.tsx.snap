// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea Component should be correctly rendered 1`] = `
<DocumentFragment>
  <div
    class="w-full"
  >
    <label
      class="block text-sm font-medium capitalize text-primary-800"
      for="mock"
    >
      Label
    </label>
    <div
      class="mt-1 rounded-md shadow-sm"
    >
      <textarea
        class="block w-full rounded-md bg-primary-50 pl-6 pr-10 text-base text-primary-900 border-primary-500 focus:border-primary-900 focus:ring-primary-900"
        data-testid="briskly-custom-textarea"
        id="mock"
        rows="5"
      />
    </div>
    <p
      class="mt-1 h-5 overflow-hidden text-ellipsis whitespace-nowrap text-sm capitalize text-error-700"
    />
  </div>
</DocumentFragment>
`;

exports[`TextArea Component should be correctly rendered when disabled 1`] = `
<DocumentFragment>
  <div
    aria-disabled="true"
    class="w-full opacity-50"
  >
    <label
      class="block text-sm font-medium capitalize text-primary-800"
      for="mock"
    >
      Label
    </label>
    <div
      class="mt-1 rounded-md shadow-sm"
    >
      <textarea
        class="block w-full rounded-md bg-primary-50 pl-6 pr-10 text-base text-primary-900 border-primary-500 focus:border-primary-900 focus:ring-primary-900"
        data-testid="briskly-custom-textarea"
        disabled=""
        id="mock"
        rows="5"
      />
    </div>
    <p
      class="mt-1 h-5 overflow-hidden text-ellipsis whitespace-nowrap text-sm capitalize text-error-700"
    />
  </div>
</DocumentFragment>
`;

exports[`TextArea Component should be correctly rendered when hasError 1`] = `
<DocumentFragment>
  <div
    class="w-full"
  >
    <label
      class="block text-sm font-medium capitalize text-primary-800"
      for="mock"
    >
      Label
    </label>
    <div
      class="mt-1 rounded-md shadow-sm"
    >
      <textarea
        class="block w-full rounded-md bg-primary-50 pl-6 pr-10 text-base text-primary-900 border-error-700 focus:border-error-700 focus:ring-error-700"
        data-testid="briskly-custom-textarea"
        id="mock"
        rows="5"
      />
    </div>
    <p
      class="mt-1 h-5 overflow-hidden text-ellipsis whitespace-nowrap text-sm capitalize text-error-700"
    >
      Mocked Error
    </p>
  </div>
</DocumentFragment>
`;
